support add --nocopy <url>

jmank88 opened the pull request:
This PR enables `add --nocopy` support for URLs, by ~setting `DagBuilderParams.URL` when `files.FileInfo.AbsPath()` is a compatible URL.~ updating dependencies.

~Draft until `go-ipfs-cmds v0.0.4` is released~
Fixes #6065

Stebalien commented on core/coreunix/add.go:
```diff
@@ -120,6 +121,11 @@ func (adder *Adder) add(reader io.Reader) (ipld.Node, error) {
+		if absPath := fi.AbsPath(); filestore.IsURL(absPath) {
+			params.URL = absPath
+		}
+	}
```
I'm not sure we need this given that we're setting AbsPath. We can probably get rid of this parameter all together and construct a `WebFile` in `core/commands/urlstore.go` instead.

jmank88 replied:
Would that cause the data to be downloaded again?

Stebalien replied:
In the `urlstore` command, we download the data on the server (I'm talking about two different commands, we can probably fix _this_ later).

jmank88 replied:
Ah I see - this code is completely redundant, and if we use a `WebFile` from `urlstore`, then the URL param is unused. Would `urlstore` itself be redundant also though? Is it still worth keeping around?

Stebalien replied:
urlstore will be redundant but we should keep it around for compatibility (for a few releases as it _is_ experimental).

jmank88 commented:
This approach has some side effects which urlstore does not (or at least doesn't log about) - all of the paths up to the root are added as directories:
```sh
> ipfs add --nocopy https://google.cloud.storage/v0/b/bucket/o/Qm...?alt=media&token=XXXXX
added Qm... https:/google.cloud.storage/v0/b/bucket/o/Qm...?alt=media
added Qm... https:/google.cloud.storage/v0/b/bucket/o
added Qm... https:/google.cloud.storage/v0/b/bucket
added Qm... https:/google.cloud.storage/v0/b
added Qm... https:/google.cloud.storage/v0
added Qm... https:/google.cloud.storage
added Qm... https:
> ipfs cat Qm...
hello world!
```
Also note the dropping of the `&token...` and the change from `https://` to `https:/` - I'm not sure what to make of that.

Stebalien commented:
The slash thing is the auto path canonicalization. Actually, I think we have to go back and fix this in go-ipfs-cmds. When we create the web file, we need to set `fpath` to something sane. Ideally the "filename" from the URL, if we can extract a reasonable one.

jmank88 commented:
I'll put up a fix for `fpath`. Then this PR is reduced to 'update the dependencies'.

Stebalien commented:
This PR will probably be folded into #6019 once we fold #6120 into that as well (batch of interdependent changes)

Stebalien commented:
(folded)
